---
title: "12SV5 reference"
output: html_notebook
---

# Setup 
```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

## Packages and functions
```{r}
library(Biostrings) 
library(here)
library(ShortRead) # for clean()
library(tidyverse)

source(here('code', 'functions', 'find_primer_pair.R'))
source(here('code', 'functions', 'query_ncbi.R'))
source(here('code', 'functions', 'ncbi_to_dada.R'))
source(here('code', 'functions', 'get_binomial.R'))
```

## Primer sequences
```{r}
# 12SV5
V5F <- DNAString('TAGAACAGGCTCCTCTAG')
V5R <- DNAString('TTAGATACCCCACTATGC')

primers <- list(V5F, V5R)
```

## Dietary animals

*2020/11/16*: Starting with minimal database for first-pass analysis of blocking primer experiment (done by Sharon October 2020)

Thoughts when clicking through sequences to assemble:

* Multiple close matches to human, chicken, turkey, goat, Bos spp.  Will exact matching accommodate these?  Do I need to tune DADA2 parameters?
* F. prau comes up!! (>90% identity)

```{r}
animals <- c('Homo sapiens', 
             'Bos taurus', 
             'Gallus gallus',
             'Meleagris gallopavo',
             'Sus scrofa',
             'Bison bison',
             'Salmo salar',
             'Oreochromis niloticus',
             'Capra hircus', 
             'Epinephelus morio', 
             'Coryphaena hippurus',
             'Gadus morhua',
             # BLAST cutoff here
             'Paralichthys lethostigma',
             'Callinectes sapidus',
             'Oncorhynchus kisutch',
             'Apis mellifera')
```

# Compile reference sequences

Rough workflow is to 
* pull sequences from curated databases (e.g. RefSeq mitochondria) first; 
* pull sequences from a general NCBI query

## RefSeq mitochondrial database

Downloaded from RefSeq FTP site on 5/18/2021.
```{r}
fs <- 
     here('data', 'raw', 'organelles', 'refseq_mito') %>% 
     list.files(full.names = TRUE)

basename(fs)

mito <- c(readDNAStringSet(fs[1]), readDNAStringSet(fs[2]))
mito
```

```{r}
head(names(mito))

# Simplify naming from current
names(mito) <- 
     names(mito) %>%
     str_split(' mitochondrion') %>% 
     lapply('[[', 1) %>% 
     unlist()

head(names(mito))
```

```{r}
# Are any names duplicated?
any(duplicated(names(mito)))
```

Write to file
```{r}
# writeXStringSet(mito, 
#                 here('data', 'processed', 'parsed-refs', 'RefSeq',
#                      'refseq_mito_species.fasta'))
```

### Look for primer binding sites

```{r}
# Note this discrepancy-- could potentially do better screening of these sequences.  Worried that I am tossing things for Ns that occur nowhere near 12SV5
length(refseq.mito)
length(clean(refseq.mito))
```

```{r}
refseq.12SV5 <- find_primer_pair(clean(mito), 
                                 fwd = primers[[1]],
                                 rev = primers[[2]])

cat(length(refseq.12SV5), 'sequences have the primer set')
```

### Subset by food species

```{r}
# Find indices of entries matching 
animals.i <- 
        lapply(animals, grep, x = names(refseq.12SV5)) %>%
        unlist()

# Subset
refseq.12SV5 <- refseq.12SV5[animals.i]
```

What foods are still missing?

```{r}
foods.refseq <- sub('\\S+\\s', '', names(refseq.12SV5))

# Subset remaining foods based on this list
animals[lapply(animals,
               function(x){length(grep(x, foods.refseq))}) == 0]
```

## NCBI nucleotide record

```{r}
# Pull sequences from NCBI
ncbi.12SV5 <- query_ncbi(marker = '12S',
                         organisms = animals)
```

This is from the total number of available sequences

```{r}
length(ncbi.12SV5)
```

Now look for primer binding sites within retrieved sequences.
Note that current strategy to remove ambiguous nucleotides is problematic-- because a single ambiguous nucleotide far from the amplicon could lead to throwing away a usable sequence
TODO: Needs to be addressed in find_primer_pair.R

```{r}
ncbi.12SV5 <- find_primer_pair(clean(ncbi.12SV5), 
                                 fwd = primers[[1]],
                                 rev = primers[[2]])

cat(length(ncbi.12SV5), 'sequences have the primer set')
```

Now we need to do some cleaning:
* Names need to be simplified
* Sequences that are the same and that come from the same species can be de-duplicated
* Sequences that are different and come from the same species must be preserved

All of this is done inside the ncbi_to_dada() function

Note here getting error that "Some IDs appear to be invalid"
Most begin with MW, MH
Looks like ~150 in all 

```{r}
epcr.12SV5.ncbi <- acc_to_dada(ncbi.12SV5)
```

## Combine results for final reference

```{r}
reference.12SV5 <- append(epcr.12SV5.refseq.foods, epcr.12SV5.ncbi)
```

```{r}
# What foods are still missing as of 2020/11/19?
present <- get_binomial(reference.12SV5)

# Subset remaining foods based on this list
animals[!(animals %in% present)]
```

Only missing crab and bee.  Leave these for now.  

### Write to file

```{r}
# assignTaxonomy format
writeXStringSet(reference.12SV5,
                here('data', 'processed', 'dada2-compatible','12SV5',
                     '20201119_Minimal 12SV5, RefSeq with NCBI patch, taxonomy.fasta'))
```

```{r}
# assignSpecies format
names(reference.12SV5) <- make.unique(present, sep = ' ')

writeXStringSet(reference.12SV5,
                here('data', 'processed', 'dada2-compatible','12SV5',
                     '20201119_Minimal 12SV5, RefSeq with NCBI patch, species.fasta'))
```

## Agastya's reference

```{r}
ref <- 
     here('data', 'processed', 'dada2-compatible', '12SV5', 
          '20210315_12SV5, AS curated, species.fasta') %>% 
     readDNAStringSet()

ref
```

```{r}
hist(width(ref))
```

```{r}
# Any duplicated?
seqs <- as.character(ref)
names(seqs) <- NULL
length(unique(seqs))
```
No. 

```{r}
# How many with Ns?
length(clean(ref))

# Remove these
ref <- clean(ref)
```

```{r}
# Think most names are okay, but not all
# Most names are formatted ACC Genus species, try extracting these with regular expression

headers <- 
     data.frame(header = names(ref)) %>% 
     separate(header, into = c('accession', 'genus', 'species'), sep = ' ')
```

```{r}
unique(headers$genus)
```

```{r}
# Collect indices of values to remove
remove <- numeric()

# pdb accessions
remove <- c(remove, 
            grep('pdb', headers$accession), # PDB values
            grep('duck61_LP-SSHFS3.50', headers$genus), # Duck virus
            grep('C.coturnix', headers$genus)) # Unclear name (quail?)
            
headers <- headers[!(1:dim(headers)[1] %in% remove), ]
```

```{r}
# Reconstruct names
headers <- mutate(headers,
                  name = paste(accession, genus, species))
```

```{r}
# Pull only those sequences with an accompanying (filtered) name
sum(names(ref) %in% headers$name)
ref <- ref[names(ref) %in% headers$name]
```

```{r}
# Have potential problem of duplicated names (non-unique-- can DADA2 accommodate?)
headers[which(duplicated(headers$accession)), ]
```

```{r}
# Write updated reference to file
writeXStringSet(ref,
                here('data', 'processed', 'dada2-compatible','12SV5',
                     '20210510, 12SV5_AS curated, species, filtered.fasta'))
```

