---
title: "Extend reference"
output: html_notebook
---

Can load in middle, no need to query if no updates to inputs!

# Setup

```{r setup, include = FALSE, echo = FALSE}
require(knitr)
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r}
library(Biostrings)
library(here)
library(rentrez)
library(tidyverse)

source(here('code', 'functions', 'find_primer_pair.R'))
```

```{r}
# Plotting themes
theme_set(
     theme_bw() +
          theme(
               axis.text = element_text(size = 12),
               axis.title.x = ggtext::element_markdown(size = 14),
               axis.title.y = ggtext::element_markdown(size = 14),
               legend.title = element_text(size = 12),
               strip.text = element_text(size = 12)
          )
)
```

# Read in data
```{r}
# Read in current version of reference
current <- 
     here('data', 
          'processed', 
          'dada2-compatible', 
          'trnL', 
          'trnLGH.fasta') %>%
     readDNAStringSet()

current
```

# QC

Common errors:

- Duplicated sequences
- Sequences with non-AGCT nucleotides

```{r}
# Organize as dataframe for referencing back to accession
current.df <- 
     data.frame(
          name = names(current),
          seq = current
     ) 

row.names(current.df) <- NULL
```

```{r}
# Separate accession number from species name
current.df <- 
     separate(current.df,
              name,
              sep = ' ',
              into = c('accession', 'species'),
              extra = 'merge',
              remove = FALSE)
```

```{r}
# Combine species name and sequence to make unique identifier
current.df <- mutate(current.df,
                     ID = paste(species, seq))
```

Now, check for errors:

```{r}
any(duplicated(current.df$ID))
```

```{r}
# Check for degenerate nucleotide characters
grep('[AGCT]*[^AGCT]+', current.df$seq)
```

# Add new sequences

## Pull sequence data

Need to do this so primers are also included.

```{r}
fs <- 
     here('data',
          'processed',
          'sql-compatible') %>% 
     list.files(pattern = 'additions.csv',
                full.names = TRUE)

basename(fs)
```

```{r}
add <- 
     sapply(fs, read_csv) %>% 
     bind_rows()

add
```
Now, can use these accession numbers to query NCBI for sequence data:
```{r}
# Search for IDs
ids <- lapply(add$accession_num, 
              entrez_search, 
              db='nucleotide', retmax = 10000, use_history=TRUE)

class(ids)
str(ids[[1]])
```

```{r}
# Formulate this as a new query
ids <- 
     lapply(ids, function(x){x$ids}) %>% 
     unlist()

ids
```

```{r}
# Use these IDs to fetch sequence data
records <- 
     entrez_fetch(db = 'nucleotide', id = ids, rettype = 'fasta')

# Returns one big character string
class(records)
length(records)
nchar(records)
```

```{r}
# Format as a DNA string set
write(records,
      here('data', 
           'processed',
           'sql-compatible',
           'fastas',
           '2021 Manual additions.fasta'))

# Read back in
records <- 
     here('data', 
           'processed',
           'sql-compatible',
           'fastas',
           '2021 Manual additions.fasta') %>% 
     readDNAStringSet()

records
```

## Trim primers

```{r}
trnLG <- DNAString('GGGCAATCCTGAGCCAA')
trnLH <- DNAString('CCATTGAGTCTCTGCACCTATC')

# Keep defaults: 20% error tolerance, no primer trimming
records.trim <- find_primer_pair(records,
                                 trnLG,
                                 trnLH)
```

## Rename
Lots of variation in descriptors.  Try pulling taxon name by NCBI query of accession number

```{r}
accs <- 
     names(records.trim) %>% 
     str_split(pattern = '\\s') %>% 
     lapply('[[', 1) %>% 
     unlist()

head(accs)
```

```{r}
# Look up accession taxonomy with metacoder package functions
# Note-- this takes a long time depending on # of sequence IDs
taxmap <- metacoder::lookup_tax_data(accs, 
                                     type = 'seq_id')
```

```{r}
str(taxmap$data)
```

```{r}
# Want to combine these two, to get name for each query id
# Get query info
query_names <- 
     data.frame(acc = taxmap$data$query_data,
                taxon_id = names(taxmap$data$query_data))

# Join to NCBI names
ncbi_names <- taxmap$data$tax_data

query_names <- 
     left_join(query_names, ncbi_names)

query_names
```

```{r}
# Any names missing?
any(is.na(query_names$ncbi_name))
```

```{r}
# How complete are entries?
query_names %>% 
     group_by(ncbi_rank) %>% 
     count()
```

```{r}
# Hmmm-- what's  no rank?
query_names %>% 
     filter(ncbi_rank == 'no rank')
```
```{r}
# Rename sequences
names(records.trim) <- paste(query_names$acc,
                             query_names$ncbi_name)
```

## Combine

Want to synchronize with the database representation of the current reference

```{r}
head(current.df)
```


```{r}
# Format as a dataframe
add.df <- 
     data.frame(name = names(records.trim),
                seq = as.character(records.trim)) %>% 
     separate(name,
              into = c('accession', 'species'),
              extra = 'merge', 
              sep = '\\s') %>% 
     mutate(ID = paste(species, seq))
```
### QC
#### Duplicates
```{r}
# Check for duplicates
# In this list:
any(duplicated(add.df$ID))

# Remove
add.df <- add.df[!duplicated(add.df$ID), ]
```

```{r}
# In the existing reference:
any(add.df$ID %in% current.df$ID)

# Remove
add.df <- add.df[!(add.df$ID %in% current.df$ID), ]
```

#### Sequences with Ns

```{r}
# Remove sequences containing Ns
any(grepl(pattern = '[AGCT]*[^AGCT]+', add.df$seq))

add.df %>% 
     filter(grepl(pattern = '[AGCT]*[^AGCT]+', seq)) 
```
Ok, basil with an R and D.  Remove:

```{r}
add.df <- 
     filter(add.df,
            !grepl(pattern = '[AGCT]*[^AGCT]+', seq))
```

```{r}
update.df <- bind_rows(current.df,
                        add.df)
```

## Save

```{r}
# Sort alphabetically (first by species name, and then accession number)
update.df <- arrange(update.df, 
                     species,
                     accession)

# Convert back to DNAStringSet object
trnL <- update.df$seq
names(trnL) <- paste(update.df$accession, update.df$species)

trnL <- DNAStringSet(trnL)
trnL
```

```{r}
# Manual edits:
# Remove suspected mislabeled curry leaf entry (identical to peanut)
trnL <- trnL[!(names(trnL) == 'JX390727.1 Murraya koenigii')]
```

```{r}
# # Save to file
# writeXStringSet(trnL,
#                 here('data',
#                      'processed',
#                      'dada2-compatible',
#                      'trnL',
#                      '20220214_trnLGH_primers included, additions.fasta'))
```

## Descriptive stats

```{r}
current
```

Add features of interest: length, GC content

```{r}
# Helper function for GC
gc <- function(seq){
     total <- nchar(seq)
     g <- str_count(seq, 'G')
     c <- str_count(seq, 'C')
     
     (g+c)/total
}
```

```{r}
current.df <- 
     current.df %>% 
     mutate(length = nchar(seq),
            gc = gc(seq))
```

```{r}
# Plot
ggplot(current.df, aes(x = length)) +
     geom_histogram(binwidth = 2,
                    boundary = 0) +
     # ylim(0, 190) +
     labs(y = 'Number of<br>database sequences',
          x = '*trnL* sequence length')
```

```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            paste0(plotID, '_trnL sequence length histogram.png')),
       height = 3, width = 3)
```

```{r}
# Plot
ggplot(current.df, aes(x = gc)) +
     geom_histogram(binwidth = 0.01,
                    boundary = 0) +
     # ylim(0, 190) +
     labs(y = 'Number of<br>database sequences',
          x = 'GC%') 
```
```{r}
plotID <- paste(Sys.Date(), as.character(sample(000:999,1)), sep=".")

ggsave(here('results',
            paste0(plotID, '_trnL GC histogram.png')),
       height = 3, width = 3)
```

```{r}
current.df
```

