---
title: "trnL reference"
author: "brianna petrone"
output: html_document
---

# Setup 
### Packages and functions
```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r}
library(Biostrings)
library(here)
library(ShortRead) # for clean()
library(tidyverse)

source(here('code', 'functions', 'find_primer_pair.R'))
source(here('code', 'functions', 'get_binomial.R'))
source(here('code', 'functions', 'query_ncbi.R'))
```

### Primer sequences
```{r}
trnLG <- DNAString('GGGCAATCCTGAGCCAA')
trnLH <- DNAString('CCATTGAGTCTCTGCACCTATC')

primers <- list(trnLG, trnLH)
```

### Dietary plants
```{r}
plants <- 
     here('data', 'raw', 'human-foods.csv') %>% 
     read.csv(stringsAsFactors = FALSE) %>%
     filter(category == 'plant') %>% 
     pull(scientific_name)

head(plants)
```

# Compile reference sequences

## RefSeq plastid database

Downloaded from RefSeq FTP site on 6/1/2021.
```{r}
fs <- 
     here('data', 'raw', 'organelles', 'refseq_plastid') %>% 
     list.files(full.names = TRUE)

basename(fs)

plastid <- 
     c(readDNAStringSet(fs[1]), 
       readDNAStringSet(fs[2]),
       readDNAStringSet(fs[3]),
       readDNAStringSet(fs[4]))

plastid
```

```{r}
head(names(plastid))

# Simplify naming from current
names(plastid) <- 
     names(plastid) %>%
     str_split(' chloroplast| plastid| apicoplast') %>% 
     lapply('[[', 1) %>% 
     unlist()

names(plastid) <- 
     names(plastid) %>%
     str_split(' voucher| isolate| culture') %>% 
     lapply('[[', 1) %>% 
     unlist()

names(plastid) <- 
     names(plastid) %>%
     str_split(' complete') %>% 
     lapply('[[', 1) %>% 
     unlist()

# May also want to remove strain, cultivar info? TBD.

head(names(plastid))
```

```{r}
# Peek at any names that aren't binomial
names(plastid)[grep('\\S+\\s\\S+\\s\\S+\\s\\S+', # More than 3 words
                    names(plastid))]
```


```{r}
# Are any names duplicated?
any(duplicated(names(plastid)))
```

```{r}
# Write to file
# writeXStringSet(plastid,
#                 here('data', 'processed', 'parsed-refs', 'RefSeq',
#                      'refseq_plastid_species.fasta'))
```

### Find primer binding sites

```{r}
# Note this discrepancy-- could potentially do better screening of these sequences.  Worried that I am tossing things for Ns that occur nowhere near trnL-P6
length(plastid)
length(clean(plastid))
```

```{r}
refseq.trnL <- find_primer_pair(clean(plastid), 
                                fwd = primers[[1]],
                                rev = primers[[2]])

cat(length(refseq.trnL), 'sequences have the primer set')
```

### Subset by food species

```{r}
# Find indices of entries matching 
plants.i <- 
     lapply(plants, grep, x = names(refseq.trnL)) %>%
     unlist()

cat('There are', length(plants), 'food plants in our query\n')

# Subset
refseq.trnL <- refseq.trnL[plants.i]
cat(length(refseq.trnL), 'have a trnL sequence in the RefSeq plastid database')
```

What foods are still missing?

```{r}
plants.refseq <- sub('\\S+\\s', '', names(refseq.trnL))

# Subset remaining foods based on this list
plants.missing <- 
     plants[lapply(plants,
                   function(x){length(grep(x, plants.refseq))}) == 0]

plants.missing
```

## NCBI

```{r}
# Pull sequences from NCBI
ncbi.trnL <- query_ncbi(marker = 'trnL',
                        organisms = plants.missing)
```

This is from the total number of available sequences

```{r}
length(ncbi.trnL)
length(clean(ncbi.trnL))
```

Now look for primer binding sites within retrieved sequences.
Note that current strategy to remove ambiguous nucleotides is problematic-- because a single ambiguous nucleotide far from the amplicon could lead to throwing away a usable sequence
TODO: Needs to be addressed in find_primer_pair.R

```{r}
ncbi.trnL <- find_primer_pair(clean(ncbi.trnL), 
                                 fwd = primers[[1]],
                                 rev = primers[[2]])

cat(length(ncbi.trnL), 'sequences have the primer set')
```

Now we need to do some cleaning:
* Names need to be simplified
* Sequences that are the same and that come from the same species can be de-duplicated
* Sequences that are different and come from the same species must be preserved

```{r}
# Inspect names
data.frame(names(ncbi.trnL))
```
```{r}
# Note some entries are marked as unverified
names(ncbi.trnL)[grepl('UNVERIFIED', names(ncbi.trnL))]
```

```{r}
# Remove these
length(ncbi.trnL)
ncbi.trnL <- ncbi.trnL[!(grepl('UNVERIFIED', names(ncbi.trnL)))]
length(ncbi.trnL)
```

```{r}
# Simplify naming from current
names(ncbi.trnL) <- 
     names(ncbi.trnL) %>%
     str_split(' chloroplast| plastid| apicoplast') %>% 
     lapply('[[', 1) %>% 
     unlist()

names(ncbi.trnL) <- 
     names(ncbi.trnL) %>%
     str_split(' voucher| isolate| culture') %>% 
     lapply('[[', 1) %>% 
     unlist()

names(ncbi.trnL) <- 
     names(ncbi.trnL) %>%
     str_split(' complete') %>% 
     lapply('[[', 1) %>% 
     unlist()
```

