---
title: "trnL reference"
author: "brianna petrone"
output: html_document
---

# Setup 
### Packages and functions
```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = file.path(getwd(), '..'))
```

```{r}
library(Biostrings)
library(here)
library(ShortRead) # for clean()
library(tidyverse)

source(here('code', 'functions', 'find_primer_pair.R'))
source(here('code', 'functions', 'get_binomial.R'))
```

### Primer sequences
```{r}
trnLG <- DNAString('GGGCAATCCTGAGCCAA')
trnLH <- DNAString('CCATTGAGTCTCTGCACCTATC')

primers <- list(trnLG, trnLH)
```

### Dietary plants
```{r}
plants <- 
     here('data', 'raw', 'human-foods.csv') %>% 
     read.csv(stringsAsFactors = FALSE) %>%
     filter(category == 'plant')

length(plants$scientific_name)
```

# Compile reference sequences

## RefSeq plastid database

Downloaded from RefSeq FTP site on 6/1/2021.
```{r}
fs <- 
     here('data', 'raw', 'organelles', 'refseq_plastid') %>% 
     list.files(full.names = TRUE)

basename(fs)

plastid <- 
     c(readDNAStringSet(fs[1]), 
       readDNAStringSet(fs[2]),
       readDNAStringSet(fs[3]),
       readDNAStringSet(fs[4]))

plastid
```

```{r}
head(names(plastid))

# Simplify naming from current
names(plastid) <- 
     names(plastid) %>%
     str_split(' chloroplast| plastid| apicoplast') %>% 
     lapply('[[', 1) %>% 
     unlist()

names(plastid) <- 
     names(plastid) %>%
     str_split(' voucher| isolate| culture') %>% 
     lapply('[[', 1) %>% 
     unlist()

# May also want to remove strain, cultivar info? TBD.

head(names(plastid))
```

```{r}
# Are any names duplicated?
any(duplicated(names(plastid)))
```

```{r}
# Write to file
# writeXStringSet(plastid,
#                 here('data', 'processed', 'parsed-refs', 'RefSeq',
#                      'refseq_plastid_species.fasta'))
```

### Find primer binding sites
```{r}
path <- here('data', 'processed', 'alignment-compatible', 'trnL')

for (i in seq_along(primers)){
     cat(names(primers)[i], sep='\n')
     varname <- paste0('epcr_', names(primers[i]))
     assign(varname,
            find_primer_pair(clean(refseq_plastid), 
                             # Remove ambiguous nucleotides prior to search (leaving 2,910 sequences)
                             # Or 120 sequences for plant foods only on 2019/9/16
                             fwd = primers[[i]][[1]],
                             rev = primers[[i]][[2]]))
            
     # Find how many sequences have the primer set
     cat(length(get(varname)), sep='\n')
     
     # Find what species in template food list are contained in that set
     # amplified <- get_binomial(get(varname))
     # cat(template_panel %>%
     #          filter(scientific_name %in% amplified) %>%
     #          pull(scientific_name) %>%
     #          as.character(), '\n')
     
     # Write to file
     # fname <- paste0(paste(Sys.Date(), 'RefSeqPlastid', 'ePCR', names(primers[i]), sep='_'), '.fasta')
     # writeFasta(get(varname), file.path(path, fname))
}
```

```{r}
trnLCD <- readDNAStringSet(file.path(path, 'ePCR_trnLCD_RefSeqPlastid.fasta'))
trnLCH <- readDNAStringSet(file.path(path, 'ePCR_trnLCH_RefSeqPlastid.fasta'))
trnLGH <- readDNAStringSet(file.path(path, 'ePCR_trnLGH_RefSeqPlastid.fasta'))
```

```{r}
source(here('code', 'functions', 'get_binomial.R'))
```
Select only foods in list from each database
```{r}
refs <- list(trnLCD, trnLCH, trnLGH)
short_names <- lapply(refs, get_binomial)

foods_only <- function(i){
     refs[[i]][short_names[[i]] %in% plants$scientific_name]
}

refs_food <- lapply(seq_along(refs), foods_only)
```
How does this change our number of sequences?
```{r}
lapply(refs, length)
```
```{r}
lapply(refs_food, length)
```
What foods are missing?
```{r}
included <- 
     lapply(refs_food, get_binomial) %>%
     unlist() %>%
     unique()

plants %>% 
     filter(!(scientific_name %in% included)) %>% 
     View()
```

```{r}
names(refs_food) <- c(paste0(Sys.Date(), '_', 'ePCR_trnLCD_foods.fasta'),
                      paste0(Sys.Date(), '_', 'ePCR_trnLCH_foods.fasta'),
                      paste0(Sys.Date(), '_', 'ePCR_trnLGH_foods.fasta'))

for (i in seq_along(refs_food)){
     writeXStringSet(refs_food[[i]], file.path(path, names(refs_food)[i]))
}
```
Try looking at alignment
```{r}
CD.algn <- AlignSeqs(refs.food[[1]])
```
```{r}
GH.algn <- AlignSeqs(refs.food[[3]])
BrowseSeqs(GH.algn)
```

